test/e2e/network/ingress_scale.go

```diff
@@ -0,0 +1,329 @@
+				}
+			}
+
+			By("Cleaning up cloud resources...")
+			framework.CleanupGCEIngressControllerWithTimeout(gceController, ingressesCleanupTimeout)
+		})
+
+		It("Create and update ingress should happen promptly with small/medium/large amount of ingresses", func() {
+			By(fmt.Sprintf("Bringing up %d backend servers...", scaleTestNumBackends))
+			testDeploy := generateScaleTestBackendDeploymentSpec(scaleTestNumBackends)
+			_, err := f.ClientSet.ExtensionsV1beta1().Deployments(ns).Create(testDeploy)
+			defer func() {
+				By(fmt.Sprintf("Deleting deployment %s...", testDeploy.Name))
+				if err := f.ClientSet.ExtensionsV1beta1().Deployments(ns).Delete(testDeploy.Name, nil); err != nil {
+					framework.Logf("Failed to delete deployment %s: %v", testDeploy.Name, err)
+				}
+			}()
+			framework.ExpectNoError(err)
+
+			By(fmt.Sprintf("Ensuring TLS secret %s...", scaleTestSecretName))
+			jig.PrepareTLSSecret(ns, scaleTestSecretName, scaleTestHostname)
+
+			currentNum := new(int)
+
+			prepareIngsFunc := func(goalNum int) {
+				var ingWg sync.WaitGroup
+				numToCreate := goalNum - *currentNum
+				ingWg.Add(numToCreate)
+				errQueue := make(chan error, numToCreate)
+				start := time.Now()
+				for ; *currentNum < goalNum; *currentNum++ {
+					suffix := fmt.Sprintf("%d", *currentNum)
+					go func() {
+						defer GinkgoRecover()
+						defer ingWg.Done()
+						acquireThread()
+						defer releaseThread()
+
+						start := time.Now()
+						svcCreated, ingCreated, err := createScaleTestServiceIngress(f.ClientSet, ns, suffix)
+						scaleTestSvcs = append(scaleTestSvcs, svcCreated)
+						scaleTestIngs = append(scaleTestIngs, ingCreated)
+						if err != nil {
+							framework.Logf("Failed to create test service and ingress: %v", err)
+							errQueue <- err
+							return
+						}
+						framework.Logf("Waiting for ingress %s to come up...", ingCreated.Name)
+						jig.WaitForGivenIngressWithTimeout(ingCreated, false, ingressCreateLatencyThreshhold)
+						elapsed := time.Since(start)
+						framework.Logf("Spent %s for ingress %s to come up", elapsed, ingCreated.Name)
+					}()
+				}
+
+				ingWg.Wait()
+				Expect(len(errQueue)).To(Equal(0), "Expect no error for creating services and ingresses")
+				elapsed := time.Since(start)
+				framework.Logf("Spent %s for %d ingresses to come up", elapsed, numToCreate)
+			}
+
+			measureCreatUpdateFunc := func() {
+				framework.Logf("Create one more ingress and wait for it to come up")
+				start := time.Now()
+				svcCreated, ingCreated, err := createScaleTestServiceIngress(f.ClientSet, ns, fmt.Sprintf("%d", *currentNum))
+				*currentNum = *currentNum + 1
+				scaleTestSvcs = append(scaleTestSvcs, svcCreated)
+				scaleTestIngs = append(scaleTestIngs, ingCreated)
+				framework.ExpectNoError(err)
+				framework.Logf("Waiting for ingress %s to come up...", ingCreated.Name)
+				jig.WaitForGivenIngressWithTimeout(ingCreated, false, ingressCreateLatencyThreshhold)
```

Comment:
Looking at this code some, it looks like it only waits until there's an IP address associated with the Ingress. Can/Should we measure how long until we're serving 200s on the address?

Reply:
In fact, this function does wait for the address to serve 200s, it happens in `PollIngressWithCert()` function, in which we poll from every defined rule/path:
https://github.com/kubernetes/kubernetes/blob/38290535d696f9dd6811044d77be6e2f38efe47f/test/e2e/framework/ingress_utils.go#L1154-L1177

Reply:
ah, okay, great! (I had followed the path of WaitForIngressAddress(), but missed the Poll... part.)